src/linear_algebra/basic.lean

```diff
@@ -2102,10 +2102,47 @@ end
 
 end field
 
+/-- Given `f : M₁ ≃ₗ[R] M₂` and `U` a submodule of `M₂`, `f.comap U` is the
+induced `linear_equiv` from `U.comap f.to_linear_map` to `U`. -/
+def comap [ring R] [add_comm_group M] [module R M]
+  [add_comm_group M₂] [module R M₂] (f : M ≃ₗ[R] M₂) (U : submodule R M₂) :
+  U.comap f.to_linear_map ≃ₗ[R] U := f.of_submodules _ _
+begin
+  ext x,
+  simp_rw [submodule.mem_map, submodule.mem_comap],
+  split; intro hx,
+  { rcases hx with ⟨y, hy, rfl⟩,
+    exact hy },
+  { refine ⟨f.inv_fun x, _⟩,
+    simpa }
+end
+
 end linear_equiv
 
 namespace submodule
 
+/-- Given `p` a submodule of the module `M` and `q` a submodule of `p`, `p.equiv_subtype_map q`
+is the natural `linear_equiv` between `q` and `q.map p.subtype`. -/
+noncomputable def equiv_subtype_map [ring R] [add_comm_group M] [module R M]
```

Comment:
Does this need to be `noncomputable`?

Reply:
I used `linear_equiv.of_bijective` which is noncomputable. The inverse function need to case on a prop so we would need to use `choice` if we construct the inverse instead

Reply:
> The inverse function need to case on a prop

Doesn't that just need a decidability assumption?

Reply:
Sorry, I misspoke. I meant to say we need to cases on something like `x \in _.map _` which is data

Reply:
I'm pretty sure you don't need to, because the only actual data this equiv produces is a term in `M`, and you know that `(equiv_subtype_map p q).symm x = ↑x`.

If you make this change, you can also relax `ring` to `semiring` etc.

Reply:
You are absolutely correct. I'll construct the inverse constructively

Reply:
Done!

Reply:
Here's most of the puzzle:
```
def equiv_subtype_map (p : submodule R M) (q : submodule R p) : q ≃ₗ[R] q.map p.subtype :=
{ to_fun := (p.subtype.dom_restrict q).cod_restrict _ (begin
    rintro ⟨x, hx⟩,
    refine ⟨x, hx, rfl⟩,
  end),
  inv_fun := λ x, ⟨⟨x, begin
    obtain ⟨x, y, h, rfl⟩ := x,
    exact y.prop,
  end⟩, begin
    obtain ⟨x, y, h, rfl⟩ := x,
    simpa using h,
  end⟩,
  map_add' := λ a b, rfl,
  map_smul' := λ c a, rfl,
  left_inv := λ x, by simp,
  right_inv := λ x, sorry }
```

Reply:
Haha, I was a bit faster :)